jgrapht-core/src/test/java/org/jgrapht/alg/CyclicTransitiveReductionTest.java

```diff
@@ -0,0 +1,323 @@
+    addEdges(graph,
+      "A", "B", "B", "C", "C", "D", "D", "A", "A", "C", "D", "B", "A", "E", "B", "G", "E", "F", "F", "G",
+      "G", "E", "E", "H", "F", "L", "H", "I", "I", "J", "J", "K", "K", "L", "L", "H", "H", "J", "I", "K",
+      "J", "N", "K", "M", "M", "N", "N", "O", "O", "P", "P", "M", "M", "O", "N", "P", "G", "M", "C", "P"
+    );
+    assertEquals(16, graph.vertexSet().size());
+    assertEquals(30, graph.edgeSet().size());
+    new CyclicTransitiveReduction<>(graph).allowSyntheticEdges(allowSyntheticEdges).reduce();
+    assertEquals(16, graph.vertexSet().size());
+    assertEquals(19, graph.edgeSet().size());
+  }
+
+  @Test
+  public void mediumCyclicalGraph2NoSynthetic() {
+    cyclicalGraphWithSingleVertexSCC(false);
+  }
+
+  @Test
+  public void mediumCyclicalGraph2Synthetic() {
+    cyclicalGraphWithSingleVertexSCC(true);
+  }
+
+  private void cyclicalGraphWithSingleVertexSCC(final boolean allowSyntheticEdges) {
+    Graph<String, DefaultEdge> graph = createEmptyGraph();
+    addVertices(graph, "A", "B", "C", "D", "E", "F");
+    addEdges(graph,
+      "A", "C", "A", "D", "A", "E", "B", "A", "C", "B", "C", "D", "C", "E", "D", "B", "D", "E", "E", "B",
+      "F", "A", "F", "B", "F", "C", "F", "D", "F", "E"
+    );
+    assertEquals(6, graph.vertexSet().size());
+    assertEquals(15, graph.edgeSet().size());
+    new CyclicTransitiveReduction<>(graph).allowSyntheticEdges(allowSyntheticEdges).reduce();
+    assertEquals(6, graph.vertexSet().size());
+    assertEquals(6, graph.edgeSet().size());
+  }
+
+  @Test
+  @Category(SlowTests.class)
+  public void randomisedGraphsWithSCCsDifferentSizesNoSynthetic() {
+    randomisedGraphsWithSCCsDifferentSizes(false);
+  }
+
+  @Test
+  @Category(SlowTests.class)
+  public void randomisedGraphsWithSCCsDifferentSizesSynthetic() {
+    randomisedGraphsWithSCCsDifferentSizes(true);
+  }
+
+  private void randomisedGraphsWithSCCsDifferentSizes(final boolean allowSyntheticEdges) {
+    Graph<String, DefaultEdge> graph;
+    long totalCTRTime = 0;
+    for (int sccCount = 3; sccCount <= 24; sccCount++) {
+      graph = createEmptyGraph();
+      int sccSize = sccCount;
+      populateGraphWithSCCs(graph, sccCount, sccSize);
+      // Shuffling vertices and edges helps increase test coverage
+      randomiseGraph(graph);
+      assertEquals(sccCount * sccSize, graph.vertexSet().size());
+      assertEquals(sccCount * naturalNumberSumGauss(sccSize - 1) + (sccCount - 1) * sccSize, graph.edgeSet().size());
+      long startTime = System.currentTimeMillis();
+      new CyclicTransitiveReduction<>(graph).allowSyntheticEdges(allowSyntheticEdges).reduce();
+      totalCTRTime += (System.currentTimeMillis() - startTime);
+      assertEquals(sccCount * sccSize, graph.vertexSet().size());
+      assertEquals(sccCount * sccCount + (sccCount - 1), graph.edgeSet().size());
+    }
+    System.out.println(
```

Comment:
Don't use println in tests.

Reply:
I generally agree that it makes sense to avoid it in most cases where it adds no value. Here in a more integrative test marked as slow, I think we can waste a bit of log file real estate to print interesting information. I put the time measurements there on purpose, because I want to be able to compare the synthetic edges case to normal mode and also make sure to only measure the sum of method execution times, not the times spent for populating and randomising the graphs. I kindly ask to make an exception here.

Reply:
FYI, after I changed the test graph generator in https://github.com/jgrapht/jgrapht/pull/1008/commits/d5f24d8f2895fadd55d2b035d2e5ea8d1da240b3, typical outputs on my workstation are:
```text
Total time cyclic transitive reduction (no synthetic edges) = 17178 ms
Total time cyclic transitive reduction (synthetic edges) = 273 ms
```
Because the difference in performance is so dramatic, I wanted to have it in the test log. If this is absolutely unacceptable, I can remove the print commands, but I find the output valuable, even though of course I know the typical test reports also contain runtime information. What do you think?

Reply:
I have now removed the print statements and intra-method timing completely in favour of a more verbose comment containing previously measured timing data in order to inform co-developers.

Reply:
Good resolution, thanks.